fix(avatar): initials are not available until image loads

juliodialpad opened the pull request:
# Fix (Avatar): Initials are not available until image loads

## :hammer_and_wrench: Type Of Change

- [x] Fix
- [ ] Feature
- [ ] Refactoring
- [ ] Documentation

## :book: Description

Changed the behavior to have the initials always displayed and put the image over so when the image is loading, initials are shown.

Fixed behavior on storybook, vue 2 version didn't detected the slot changes properly.

## :bulb: Context

Even after passing the initials, initials are not shown untill the image loads, this is replicable on storybook.

## :pencil: Checklist

- [x] I have reviewed my changes
- [x] I have added tests
- [x] I have considered the performance impact of my change

## :camera: Screenshots / GIFs


Uploading Screen Recording 2023-03-07 at 9.10.59 p.m..mov…

juliodialpad committed:
fix(avatar): initials are not available until image loads

braddialpad commented on components/avatar/avatar.vue:
```diff
@@ -203,20 +204,26 @@ export default {
   methods: {
     init () {
-      const firstChild = this.$refs.canvas.firstChild;
+      this.defaultSlotContent = this.$slots.default ? this.$slots.default[0] : null;
+      const firstChild = this.defaultSlotContent?.elm ? this.defaultSlotContent.elm : this.defaultSlotContent;
+      this.formatInitials(this.initials);
```
Why was the original firstChild method with $refs.canvas not working here? seems a bit like a more complicated way of doing the same thing, but I'm sure you had a reason for it.

juliodialpad replied:
This wasn't working correctly because the $ref.canvas content was changing with the 'kind' property, sometimes overriding the checks I was doing later so the component didn't detected the changes properly.

This way, we're checking the actual default slot content property to decide what we should render.

braddialpad replied:
as I said, knew you had a good reason 😄

braddialpad commented on components/avatar/avatar.vue:
```diff
@@ -12,7 +12,7 @@
       <slot v-if="showDefaultSlot" />
       <span
-        v-else-if="showInitials"
+        class="d-ps-absolute d-zi-base"
```
I think zi-base may cause a problem if we render within something that doesn't have a default z-index, like a popover for example. Let's try this to make sure.

juliodialpad replied:
Sure, I was thinking the same, I just wanted to have the image over the initials because the initials are always being displayed so when the image is loading, the initials are behind.

 But I'll need to test several cases to check that z-index would work correctly in all of them.

braddialpad replied:
Might work if you just move the slot after the initials in the DOM order.

juliodialpad replied:
Doesn't work because initials have `d-ps-absolute`.

Tested in popover and modal and everything works fine so we shouldn't have any problems.

iropolo commented on components/avatar/avatar.vue:
```diff
@@ -203,20 +204,26 @@ export default {
-      this.slottedInitials = this.$slots.default[0].text || this.$slots.default[0].textContent;
-      this.formatInitials(this.slottedInitials);
-    }
+    this.defaultSlotContent = this.$slots.default ? this.$slots.default[0] : null;
```
Would replace it with `this.defaultSlotContent = this.$slots.default?.[0] ? this.$slots.default[0] : null;`

iropolo replied:
Mhmmm... `this.defaultSlotContent = this.$slots.default?.[0] ?? null;` is even better

juliodialpad replied:
I didn't knew I could do it that way, thanks!

braddialpad reviewed (commented):
This component is starting to feel a bit overcomplicated, though I don't think we can fix some of the issues without changing the public interface and making a breaking change. Was thinking it may simplify things if we used props instead of slots for some of this stuff since accessing the markup within the slots is fairly cumbersome and has reactivity issues.

iropolo commented on components/avatar/avatar.vue:
```diff
@@ -308,8 +324,8 @@ export default {
     validateImageAttrsPresence () {
-      const isSrcMissing = !this.$refs.canvas.firstChild.getAttribute('src');
-      const isAltMissing = !this.$refs.canvas.firstChild.getAttribute('alt');
+      const isSrcMissing = !this.$slots.default[0].elm.getAttribute('src');
+      const isAltMissing = !this.$slots.default[0].elm.getAttribute('alt');
```
Mhm... If we are using our local `defaultSlotContent` and also updating it in `updated()` hook... why we are accessing directly to the $slots here?, could we use `defaultSlotContent` ...? Or Im missing sth...

Just asking I didnt switch the branch

juliodialpad replied:
Right, I missed that one, nice catch.

juliodialpad commented:
> This component is starting to feel a bit overcomplicated, though I don't think we can fix some of the issues without changing the public interface and making a breaking change. Was thinking it may simplify things if we used props instead of slots for some of this stuff since accessing the markup within the slots is fairly cumbersome and has reactivity issues.

You bet, this component is too complicated now IMO, but yeah, making a change in the public interface would be a breaking change and a large migration.

Automatically detecting the slot content seem good at the beginning but now its hard to maintain this component.

iropolo reviewed (approved):
LGTM, I'm agree also to start to rethink this component in the future.

juliodialpad committed:
fixed found issues

juliodialpad committed:
fixed tests

juliodialpad commented:
Fixed issues I found while testing avatar over popover and modal to check that `d-zi-base` didn't break anything.

Fixed tests as well.

juliodialpad committed:
fix(avatar): error while empty icon selected

juliodialpad committed:
add debounce util

juliodialpad committed:
simplify functions and tests

juliodialpad committed:
update .esplint.rec.json

juliodialpad committed:
cleanup

juliodialpad commented:
While working on Vue 3 version, found a lot of problems with the way I made it on Vue 2 and realized that there was a better way of handling that logic.

Refactored to keep almost the same logic in both Vue 2 and Vue 3 versions.

It should be cleaner to understand what's happening and also fixed some issues I found.

The z-index should work fine as the [stacking-context](https://developer.mozilla.org/en-US/docs/Web/CSS/CSS_Positioning/Understanding_z_index/The_stacking_context) of the avatar contains it.

Please give another review @braddialpad @iropolo

braddialpad reviewed (approved):
This is cleaner, nice job.
